### ISSUE TITLE
Revocable memory overcommit using reserved pool

### ISSUE THREAD

**Issue by findepi** [prestodb/presto#10512]
Today, when spill is enabled, query can allocate more memory than `query.max-memory-per-node`, provided that everything above this limit is a revocable memory.
This works well when a cluster is configured to handle concurrent workload.

## Scenario: concurrent cluster

Now, consider a cluster configured for concurrent queries, e.g.
- 40% goes to system pool (or 30%, as above)
- `query.max-memory-per-node` will be smaller, say 10% of the heap
- then reserved pool is 10% of the heap and user pool is only 50% of the heap

In such scenario, memory revocation kicks in -- and some queries starts spilling -- as soon as they allocate `memoryRevokingThreshold * user pool`, i.e. 45% of the heap. However, 10% of the heap (the reserved pool) is constantly wasted, never used.

## Scenario: single query cluster
Consider a cluster is tuned to handle single query workload. Such a cluster with have memory pools breakdown like this:
- 40% goes to system pool (or 30% is "heap headroom" when https://github.com/prestodb/presto/pull/10491 is merged)
- `query.max-memory-per-node` can be set to some big value below remaining 60% of the heap, let's say 50% of the heap
- then reserved pool is 50% of the heap and user pool is only 10% of the heap

In such scenario, memory revocation kicks in -- and query starts spilling -- as soon as it allocates `memoryRevokingThreshold * user pool`, i.e. only 9% of the heap. 
At the same time, with spill disabled, when query doesn't use revocable memory, query would be allowed to use 50% of the heap before being killed.

## Proposed solution

We could utilize resources more effectively by leveraging revocable memory's promise more. Since query allocating revocable memory makes a promise that it will give the memory back when it's needed, we could let revocable memory to allocate from both: user pool and reserved pool (actually allocating from reserved pool first).

- let query revocable memory be allocated from two pools, from reserved pool first (until reserved pool is filled up to `memoryRevokingThreshold`)
- when a query is elevated to reserved pool, all revocable memory allocated from that pool needs to be revoced
  - the operators that hold onto that memory need to get a chance to run until memory is revoked. Probably it's OK to let whole Drivers run until memory is revoked. 

cc @dain @sopel39 @ilfrin

**Comment by dain** [comment]
@nezihyigitbasi has a pending PR to merge system and user pools.  Does that change this?

**Comment by findepi** [comment]
I think it doesn't, but @nezihyigitbasi will know better.

**Comment by dain** [comment]
If spilling is enabled, why do we need the reserved pool?

**Comment by findepi** [comment]
Good question. I think the answer is twofold: first, spilling can be enabled on per-query basis and memory management is global. If reserved pool is needed for some queries, it's must stay. Second, not everything is spillable (not every operator, not every piece of memory, not in every case, etc.), so we still exhausted memory even if we spilled everything possible.

**Comment by dain** [comment]
This seems like a good plan to me

**Comment by findepi** [comment]
@nezihyigitbasi consider:
1. user pool has 10M, reserved has 10M
1. Query A allocates 7M revocable and 5M non-revocable
1. Query B allocates 5M non-revocable
1. now user pool is exhausted; let's assume Query A and B try to allocate one byte each
1. we want to promote (move) a query to the reserved pool; whichever query we take, pool is going to have 12M used out of 10M (error), unless we do revocation first

TL;DR this requires changes (including some synchronization) in the mechanism moving queries to the reserved pool.

**Comment by sopel39** [comment]
I wonder if we couldn't use similar approach (e.g: use reserved memory first) when `resource overcommit` is set to `true`.

Generally, it seem that we could utilize reserved memory pool better (and cluster memory for that matter) always by allowing at one arbitrary query (preferably the largest one) to run there at any given time. In such case we would still deliver promise that queries can finish if they are blocked on memory requests.
